libs/header/sync/sync_head.go

```diff
@@ -10,160 +10,131 @@ import (
+	// NOTE: we avoid validation as the head expired to prevent possibility of the Long-Range Attack
+	s.setSubjectiveHead(ctx, trustHead)
+	switch {
+	default:
+		log.Infow("subjective initialization finished", "height", trustHead.Height())
+		return trustHead, nil
+	case isExpired(trustHead, s.Params.TrustingPeriod):
+		log.Warnw("subjective initialization with an expired header", "height", trustHead.Height())
+	case !isRecent(trustHead, s.Params.blockTime):
+		log.Warnw("subjective initialization with an old header", "height", trustHead.Height())
+	}
+	log.Warn("trusted peer is out of sync")
+	return trustHead, nil
+}
+
+// setSubjectiveHead takes already validated head and sets it as the new sync target.
+func (s *Syncer[H]) setSubjectiveHead(ctx context.Context, netHead H) {
+	// TODO(@Wondertan): Right now, we can only store adjacent headers, instead we should:
+	//  * Allow storing any valid header here in Store
+	//  * Remove ErrNonAdjacent
+	//  * Remove writeHead from the canonical store implementation
+	err := s.storeHeaders(ctx, netHead)
 	var nonAdj *header.ErrNonAdjacent
-	if errors.As(err, &nonAdj) {
-		// not adjacent, maybe we've missed a few headers or its from the past
-		log.Debugw("attempted to append gossiped non-adjacent header", "store head",
-			nonAdj.Head, "attempted", nonAdj.Attempted)
-	} else {
-		var verErr *header.VerifyError
-		if errors.As(err, &verErr) {
-			return pubsub.ValidationReject
-		}
+	if err != nil && !errors.As(err, &nonAdj) {
 		// might be a storage error or something else, but we can still try to continue processing netHead
-		log.Errorw("appending network header",
+		log.Errorw("storing new network header",
 			"height", netHead.Height(),
 			"hash", netHead.Hash().String(),
 			"err", err)
 	}
-	// try as new head
-	return s.newNetHead(ctx, netHead, false)
-}
 
-// newNetHead sets the network header as the new subjective head with preceding validation(per
-// request).
-func (s *Syncer[H]) newNetHead(ctx context.Context, netHead H, trust bool) pubsub.ValidationResult {
-	// validate netHead against subjective head
-	if !trust {
-		if res := s.validate(ctx, netHead); res != pubsub.ValidationAccept {
-			// netHead was either ignored or rejected
-			return res
-		}
+	storeHead, err := s.store.Head(ctx)
+	if err == nil && storeHead.Height() >= netHead.Height() {
+		// we already synced it up - do nothing
+		return
 	}
 	// and if valid, set it as new subjective head
 	s.pending.Add(netHead)
 	s.wantSync()
 	log.Infow("new network head", "height", netHead.Height(), "hash", netHead.Hash())
-	return pubsub.ValidationAccept
+	return
+}
+
+// incomingNetworkHead processes new gossiped network headers.
+func (s *Syncer[H]) incomingNetworkHead(ctx context.Context, netHead H) pubsub.ValidationResult {
+	// first of all, check the validity of the netHead
+	res := s.validateHead(ctx, netHead)
+	if res != pubsub.ValidationAccept {
+		return res
+	}
```

Comment:
Now since validating every incoming header, instead of Store. So we always have one verification codepath that covers all the cases.